### ISSUE TITLE
Get rid of the circular dependency between root and client in twisted.names

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6095]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#6095|
|Type|enhancement|
|Created|2012-10-10 18:51:53Z|
|Branch|https://github.com/twisted/twisted/tree/root-resolver-argument-6095-2|

`twisted.names.client` uses `twisted.names.root` to construct the object returned by `createResolver` (sometimes).  `twisted.names.root` uses `twisted.names.client` to bootstrap itself.

This circularity could be removed by giving `twisted.names.root.Resolver` a resolver argument so it doesn't need to import `twisted.names.client` to create one.

Attachments:

* [6095_v1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/936/93682a5bf3307df0fa368d121550135c3c4ff2db/8a7d825cd6a8b826dff75ebc50a48ad593d56fef.diff) (3106 bytes) - added by borko on 2013-01-03 21:19:28Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__6095 6095
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_root_resolver_argument_6095_2 branches/root-resolver-argument-6095-2
branch_author__rwall rwall
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1349895113000000 1349895113000000
changetime__1384617015000000 1384617015000000
version__None None
owner__rwall rwall
cc__borko
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="borko's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|borko commented|
|-|-|

I guess it's obsolate. root imports only dns, common, error from twisted.names. Cannot see client. So I think it should be closed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

See http://twistedmatrix.com/trac/browser/trunk/twisted/names/root.py?rev=36682#L115

**Comment by twisted-trac** [comment]
|<img alt="borko's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|borko commented|
|-|-|

pls give an idea of easiest unit testing this one.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to borko|
|-|-|

As long as all the code is exercised by existing unit tests, this kind of change doesn't require new unit tests (perhaps partly because it is too difficult to test, and we lack the tools to do so, but also because the unit tests exist to demonstrate the code _works_, not that it has a particular shape - any shape that makes the unit tests pass is fine, as far as the tests go; and then as developers we can apply additional constraints like "shorter is better than longer" or "easier to read is better than harder to read").

You can use [coverage.py](http://pypi.python.org/pypi/coverage) to verify that the changed code really is exercised by the unit tests (or discover that it is not).

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_6"></a>|@wallrj commented|
|-|-|

(In [39243]) Branching to 'root-resolver-argument-6095'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_7"></a>|@wallrj commented|
|-|-|

(In [39244]) Apply 6095_v1.diff from borko. Refs [#6095](https://github.com/twisted/twisted/issues/6095)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_8"></a>|@wallrj commented|
|-|-|

(In [39320]) prototype FakeResolver for testing root.Resolver(resolver) and other dns client code. refs [#6095](https://github.com/twisted/twisted/issues/6095), [#6379](https://github.com/twisted/twisted/issues/6379)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_9"></a>|@wallrj removed owner|
|-|-|

Ready for review in log:branches/root-resolver-argument-6095
 1. I reviewed Borko's 6095_v1.diff which added a new positional
    "resolver" instance argument to root.Resolver. There are a few
    problems I see with that approach:
    1. root.bootstrap is public I don't think we can add a new
       mandatory positional argument. It has to be optional.
    2. I think its more flexible to be able to provide a factory
       function which is called by root.Resolver._query to build a
       client.Resolver (or some other Resolver instance)
    3. Borko, please comment if you have any counter arguments.

 2. So I went ahead and implemented it that way.
    1. Also added some tests for root.bootstrap (there weren't any before)
    2. Build Results
       1. http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/root-resolver-argument-6095

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_10"></a>|@tomprince commented|
|-|-|

I wonder if it would be appropriate to have an actual interface to represent the factory, rather than just a callable. One thing that confused me when I first started looking is what the servers argument.

Also, the interface the factory needs to return isn't `IResolver`, since it needs `queryUDP` (which is currently only defined on `client.Resolver`).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @wallrj<br>@exarkun set status to `new`|
|-|-|

Thanks!  Sorry about the long review turn-around.

  1. twisted/names/root.py
    1. Let's think about deprecating some stuff here (separate tickets, of course).  To completely remove the circular dependency we need to stop providing a default value for the `resolverFactory` parameter.  However, `bootstrap` itself might be worth deprecating and replacing.  Mainly because of the `DeferredResolver` return type.  This should really be a `Deferred` that fires with an `IResolver` instead.
    1. I like Tom's idea of having an explicit interface for `resolverFactory`.  I see no problem with doing this as a follow-up ticket.
    1. Thanks for the doc cleanups here!
    1. `L{callable}` is usually not what's meant though - I'd just write `callable`
  1. twisted/names/test/test_rootresolve.py
    1. The `pass` in the body of `ResolverFactoryArguments` is superfluous.
    1. In `test_resolverFactoryOnlyExpectedArguments` there's an assertion that depends on the order of the elements in the result of `dict.keys`.  This should be made order-independent (`set` and `sorted` are handy - but maybe the assertion should just be against the dictionary directly).
    1. Strictly speaking I think that the reactor used in `test_resolverFactorySuppliedReactor` is a _dummy_ rather than a _stub_.  Those are both specific kinds of _fake_ in case you don't want to bother being so precise. :)  A dummy is an object that doesn't do anything.  A stub is a minimal (usually no-op) implementation of an interface.
    1. I like having separate test cases for different units being tested - like `BootstrapTests` - because it makes naming easier.  I think `test_bootstrapReturnsDeferredResolver` is redundant because it's part of `BootstrapTests`.  `test_returnsDeferredResolver` is just as unambiguous and less of a mouthful. :)
    1. In `test_bootstrapContinuesWhenAllRootHintsFail` you might want to use `TestCase.flushLoggedErrors` instead of the `addErrback` trick.
    1. There's a minor whitespace issue or two, eg `((s,),{})`.
  1. There's supposedly a new twistedchecker warning but it looks spurious to me.
  1. [#6622](https://github.com/twisted/twisted/issues/6622) isn't exactly a duplicate of this ticket but it seems like it will be significantly affected by this work.  If so it's probably worth leaving a comment there about what direction now makes sense.
  1. Some parts of Twisted Names have been ported to Python 3.  Not this part, apparently.  Think about whether the `L{str}` in the documentation should be `L{bytes}`, though, or whether that should wait until someone actually ports this module.

Thanks!  Please merge after addressing these issues.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_13"></a>|@wallrj commented|
|-|-|

(In [40663]) Branching to 'root-resolver-argument-6095-2'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_14"></a>|@wallrj set status to `assigned`|
|-|-|

Replying to [exarkun](#note_12):
> Thanks!  Sorry about the long review turn-around.
>   1. twisted/names/root.py
>     1. Let's think about deprecating some stuff here (separate tickets, of
>     course).  To completely remove the circular dependency we need to stop
>     providing a default value for the `resolverFactory` parameter.  However,
>     `bootstrap` itself might be worth deprecating and replacing.  Mainly
>     because of the `DeferredResolver` return type.  This should really be a
>     `Deferred` that fires with an `IResolver` instead.

[#6822](https://github.com/twisted/twisted/issues/6822)

>     1. I like Tom's idea of having an explicit interface for
>     `resolverFactory`.  I see no problem with doing this as a follow-up
>     ticket.

[#6823](https://github.com/twisted/twisted/issues/6823)

>     1. Thanks for the doc cleanups here!
>     1. `L{callable}` is usually not what's meant though - I'd just write `callable`

r40657

>   1. twisted/names/test/test_rootresolve.py
>     1. The `pass` in the body of `ResolverFactoryArguments` is superfluous.

r40658

>     1. In `test_resolverFactoryOnlyExpectedArguments` there's an assertion
>     that depends on the order of the elements in the result of `dict.keys`.
>     This should be made order-independent (`set` and `sorted` are handy - but
>     maybe the assertion should just be against the dictionary directly).
>     1. Strictly speaking I think that the reactor used in
>     `test_resolverFactorySuppliedReactor` is a _dummy_ rather than a
>     _stub_.  Those are both specific kinds of _fake_ in case you don't
>     want to bother being so precise. :) A dummy is an object that doesn't do
>     anything.  A stub is a minimal (usually no-op) implementation of an
>     interface.

r40659

>     1. I like having separate test cases for different units being tested -
>     like `BootstrapTests` - because it makes naming easier.  I think
>     `test_bootstrapReturnsDeferredResolver` is redundant because it's part of
>     `BootstrapTests`.  `test_returnsDeferredResolver` is just as unambiguous
>     and less of a mouthful. :)

r40660

>     1. In `test_bootstrapContinuesWhenAllRootHintsFail` you might want to use
>     `TestCase.flushLoggedErrors` instead of the `addErrback` trick.

r40662

>     1. There's a minor whitespace issue or two, eg `((s,),{})`.

r40661.

>   1. There's supposedly a new twistedchecker warning but it looks spurious to
>   me.

Yep. Spurious. My branches always get that error. Not sure why?

>   1. [#6622](https://github.com/twisted/twisted/issues/6622) isn't exactly a duplicate of this ticket but it seems like it will
>   be significantly affected by this work.  If so it's probably worth leaving a
>   comment there about what direction now makes sense.

ticket:6622#comment:13

>   1. Some parts of Twisted Names have been ported to Python 3.  Not this part,
>   apparently.  Think about whether the `L{str}` in the documentation should be
>   `L{bytes}`, though, or whether that should wait until someone actually ports
>   this module.

I decided not do that here. Too tired and the Python3 builder is offline anyway.

Thanks again for the review.

I'll merge as soon as the builders are back online and I get some clean builds.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_15"></a>|@wallrj set status to `closed`|
|-|-|

(In [40770]) Merge root-resolver-argument-6095-2: Remove circular dependency between root and client

Authors: borko, rwall
Reviewers: tom.prince, exarkun
Fixes: [#6095](https://github.com/twisted/twisted/issues/6095)
Refs: [#6622](https://github.com/twisted/twisted/issues/6622), [[#6822](https://github.com/twisted/twisted/issues/6822)](https://github.com/twisted/twisted/issues/6822), [#6823](https://github.com/twisted/twisted/issues/6823)

Partially removed the circular dependency between twisted.names.root and
twisted.names.client. #6822 will be the next step in completely removing the
dependency.

In doing this, twisted.names.root.Resolver now accepts a resolverFactory
argument, which makes it possible to control how root.Resolver performs
iterative queries to authoritative nameservers.

Also added test coverage for twisted.names.root.bootstrap.

Also added and updated various API documentation in root.
